### ISSUE TITLE
Support rubocop:push/:pop in source

### ISSUE THREAD

**Issue by steakknife** [bbatsov/rubocop#1458]
### Generic example of current usage 

```ruby
# some.rb


# ...

# rubocop:disable Some/Cop
# Common-sense violation of Some/Cop
# rubocop:enable Some/Cop

# From here on, Some/Cop is always enabled, regardless of rubocop config.
# ...

```

### Additionally, it would be awesome to have

```ruby
# ...

# rubocop:push # Some/Cop <-- some optional comment
# rubocop:disable Some/Cop
# Common-sense violation of Some/Cop
# rubocop:pop # Some/Cop <-- some optional comment

# From here on, cops are what they were.
# ...
```

### See also: clang

```c
#pragma clang diagnostic push
#pragma clang diagnostic ... // look the other way

// naughty bits here

#pragma clang diagnostic pop
```

(Thanks for publishing `rubocop`, bribery here: {:cake: | :moneybag:}+ )

**Comment by bbatsov** [comment]
It's not totally clear to me what's the problem that has to be solved. E.g. why would you disable a cop, which is not enabled?

**Comment by alexdowad** [comment]
I suggest this can be closed.

**Comment by mruzicka** [comment]
This is a useful feature, please reopen.

The use case is:
You want to disable certain cop for a pice of code which for some reason doesn't satisfy it and then restore the cop's enable/disable status to its original state before you disabled it without having to _care_ what the original state of the cop was. Perhaps it was disabled in the configuration, in which case using `#rubocop:enable` after the violating piece of the code would erroneously enable it.

**Comment by jonas054** [comment]
> Perhaps it was disabled in the configuration, in which case using #rubocop:enable after the violating piece of the code would erroneously enable it.

No, what happens when you have a cop disabled in configuration and then disabled again in a comment is that you get a `Lint/UnneededCopDisableDirective` offense. A `rubocop:enable` comment only cancels a preceding `rubocop:disable`.

When we added the `Lint/UnneededCopDisableDirective` cop, we chose a path that's incompatible with push/pop functionality. We do care about the original state of the cop.
